### PR TITLE
KAFKA-13767: Fetch from consumers should return immediately when preferred read replica is defined by the leader

### PR DESCRIPTION
**Pull request by bozhao12**
When a replica selector is configured, the partition leader computes a preferred read replica for any fetch from the consumers. When the preferred read replica is not the leader, the leader returns the preferred read replica with `FetchDataInfo(LogOffsetMetadata.UnknownOffsetMetadata, MemoryRecords.EMPTY)` to the `ReplicaManager`. This causes the fetch to go into in the fetch purgatory because the exit conditions are not met. In turns out that the delayed fetch is not completed until the timeout is reached because the delayed fetch ignores partition with an unknown offset (-1). If the fetch contains only one partition, the fetch is unnecessarily delayed by the timeout time (500ms by default) to only inform the consumer that it has to read from a follower.

This patch fixes the issue by completing the fetch request immediately when a preferred read replica is defined.

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 1]
```diff
@@ -1028,6 +1028,7 @@ class ReplicaManager(val config: KafkaConfig,
     var bytesReadable: Long = 0
     var errorReadingData = false
     var hasDivergingEpoch = false
+    var notLocalPreferredReadReplica = false
```
Comment (dajac) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: nit: `hasPreferredReadReplica` might be a better name. What do you think?

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 2]
```diff
@@ -1037,6 +1038,8 @@ class ReplicaManager(val config: KafkaConfig,
         errorReadingData = true
       if (logReadResult.divergingEpoch.nonEmpty)
         hasDivergingEpoch = true
+      if (logReadResult.preferredReadReplica.getOrElse(config.brokerId) != config.brokerId)
```
Comment (dajac) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: I think that we could use `logReadResult.preferredReadReplica.nonEmpty` because it is only set when a follower as been selected.

File: core/src/main/scala/kafka/server/ReplicaManager.scala [hunk 3]
```diff
@@ -1046,7 +1049,9 @@ class ReplicaManager(val config: KafkaConfig,
     //                        3) has enough data to respond
     //                        4) some error happens while reading data
     //                        5) we found a diverging epoch
-    if (timeout <= 0 || fetchInfos.isEmpty || bytesReadable >= fetchMinBytes || errorReadingData || hasDivergingEpoch) {
+    //                        6) the preferred read replica not local replica
```
Comment (dajac) [inline core/src/main/scala/kafka/server/ReplicaManager.scala]: nit: `has a preferred read replica`?

**Review by dajac — COMMENTED** [review]
@bozhao12 Thanks for the PR. That seems to be a reasonable improvement. Could you add a unit test for the change?

I am curious... How have you found out about this and what was the impact?

I can think of two cases:
1) The consumer has caught up so it has nothing to fetch. In this case, the fetch would be delayed until the wait timeout is reached. Obviously, if the leader does not have data, the follower won't have any data neither.
2) The consumer has not caught up yet. In this case, the fetch goes into to purgatory. However the delayed fetch will be completed immediately because min bytes is likely reached. Delaying the fetch is not necessary in this case.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 1 above)
Comment (bozhao12) [reply core/src/main/scala/kafka/server/ReplicaManager.scala:1031]: There is no impact at present. I need to use the replica read capability to do some development, so I added a little log to readFromLog() to verify the correctness. I found that every time the leader elects a non-leader replica, it will trigger the fetch response and return with a delay of 500 ms , even though I'm consuming historical data. Triggered my log to print, and then traced this case

Comment (bozhao12) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: There is no impact at present. 
I need to use the replica read capability to do some development, so I added a little log to readFromLog() to verify the correctness. I found that every time the leader broker selects a non-leader replica, it will trigger the fetch response  return with a delay of 500 ms , even though I'm consuming historical data. Triggered my log to print, and then traced this case
I  will make some code modifications and add unit tests according to your suggestion.

Comment (dajac) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Interesting.. I thought that the delayed fetch would return immediately in this case. Looking at the code, it does not because the fetch offset is set to `UnknownOffsetMetadata` so the delayed fetch just ignore it. I think that your patch makes total sense. Let me know when the comments are addressed.

Comment (bozhao12) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: @dajac added a unit test and made some code modifications.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 2 above)
Comment (bozhao12) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: done

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 3 above)
Comment (bozhao12) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: done

File: core/src/test/scala/integration/kafka/server/DelayedFetchTest.scala [hunk 4]
```diff
@@ -166,6 +171,34 @@ class DelayedFetchTest {
     assertTrue(fetchResultOpt.isDefined)
   }
 
+
+  @Test
+  def testHasPreferredReadReplica(): Unit = {
```
Comment (dajac) [inline core/src/test/scala/integration/kafka/server/DelayedFetchTest.scala]: It seems that we don't really test the changes made in `ReplicaManager.fetchMessages` method here, isn't it? Overall, I think that the unit test should be in `ReplicaManagerTest` and it should ensure that a fetch does not go to the fetch purgatory if it has a preferred read replica. We might consider update `testPreferredReplicaAsLeader` directly. That test actually works because it fetches with min bytes equals to zero. If we change this to 1, it fails right away.

**Review by dajac — COMMENTED** [review]
@bozhao12 Thanks for the update. I left a comment below.

File: core/src/main/scala/kafka/server/ReplicaManager.scala (hunk 3 above)
Comment (bozhao12) [reply core/src/main/scala/kafka/server/ReplicaManager.scala]: Following your suggestion, I modified the unit test
I also found some problems when modifying the unit test, which I think are bugs

File: core/src/test/scala/integration/kafka/server/DelayedFetchTest.scala (hunk 4 above)
Comment (bozhao12) [reply core/src/test/scala/integration/kafka/server/DelayedFetchTest.scala]: @dajac  Following your  suggestion, I have modified the unit test.
I also found a problem when modifying the unit test, which I think it's a bug:
If `KafkaConfig.ReplicaSelectorClassProp`  is not configured, `ReplicaManager.replicaSelectorOpt` will  be a None object, which will causes some unit tests of preferred read replicas to not take test purpose., eg `ReplicaManagerTest.testReplicaSelector.` It just return None , didn't really execute the replica selector.
Should I mention another PR  or fix it in this PR

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 5]
```diff
@@ -1253,9 +1253,11 @@ class ReplicaManagerTest {
     val countDownLatch = new CountDownLatch(1)
 
     // Prepare the mocked components for the test
+    val props = new Properties()
+    props.put(KafkaConfig.ReplicaSelectorClassProp, "org.apache.kafka.common.replica.RackAwareReplicaSelector")
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Hum... It seems that my suggestion to update this test was perhaps not good when I see the diff. I misread the test. It might be better to keep it and add another one which is somewhat similar. Sorry for the confusion.

Comment (bozhao12) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: @dajac Thanks a lot for your suggestion, I added a new unit test.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 6]
```diff
@@ -1300,6 +1300,67 @@ class ReplicaManagerTest {
...
+    val leaderEpoch = 1
+    val leaderEpochIncrement = 2
+    val countDownLatch = new CountDownLatch(1)
+
+    // Prepare the mocked components for the test
+    val props = new Properties()
+    props.put(KafkaConfig.ReplicaSelectorClassProp, "org.apache.kafka.common.replica.RackAwareReplicaSelector")
+    val (replicaManager, _) = prepareReplicaManagerAndLogManager(new MockTimer(time),
+      topicPartition, leaderEpoch + leaderEpochIncrement, followerBrokerId,
+      leaderBrokerId, countDownLatch, expectTruncation = true, topicId = Some(topicId), extraProps = props)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Using this method create a lot of unnecessary noise in my opinion. Nowadays, we tend to use `setupReplicaManagerWithMockedPurgatories` which is simpler. `setupReplicaManagerWithMockedPurgatories` does not support racks though. I suppose that we could it accept an optional Map from broker id to rack. Could we try using this one instead? There are many example in this file.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 7]
```diff
@@ -1300,6 +1300,67 @@ class ReplicaManagerTest {
...
+
+    try {
+      val brokerList = Seq[Integer](0, 1).asJava
+
+      val tp0 = new TopicPartition(topic, 0)
+      val tidp0 = new TopicIdPartition(topicId, tp0)
+
+      initializeLogAndTopicId(replicaManager, tp0, topicId)
+
+      // Make this replica the follower
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: I suppose the replica is a leader, no?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 8]
```diff
@@ -1300,6 +1300,67 @@ class ReplicaManagerTest {
...
+          .setLeader(0)
+          .setLeaderEpoch(1)
+          .setIsr(brokerList)
+          .setZkVersion(0)
+          .setReplicas(brokerList)
+          .setIsNew(false)).asJava,
+        Collections.singletonMap(topic, topicId),
+        Set(new Node(0, "host1", 0), new Node(1, "host2", 1)).asJava).build()
+      replicaManager.becomeLeaderOrFollower(1, leaderAndIsrRequest2, (_, _) => ())
+      replicaManager.getPartitionOrException(tp0).updateFollowerFetchState(1, new LogOffsetMetadata(0), 0, 0, 0)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Why do we need this? Does the replica selector ignore the replica if it is not caught-up? It might be worth adding a small comment.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 9]
```diff
@@ -1300,6 +1300,67 @@ class ReplicaManagerTest {
...
+
+      val metadata: ClientMetadata = new DefaultClientMetadata("rack-a", "client-id",
+        InetAddress.getByName("localhost"), KafkaPrincipal.ANONYMOUS, "default")
+
+      val consumerResult = fetchAsConsumer(replicaManager, tidp0,
+        new PartitionData(Uuid.ZERO_UUID, 0, 0, 100000, Optional.empty()), minBytes = 1,
+        clientMetadata = Some(metadata))
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Could we also explicitly set the timeout? I would also add a small comment explaining why we set min bytes and max wait time.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 10]
```diff
@@ -1300,6 +1300,67 @@ class ReplicaManagerTest {
...
+
+      // Fetch from follower succeeds
+      assertTrue(consumerResult.isFired)
+
+      // Returns a preferred replica (should just be the leader, which is None)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: I suppose that the comment is not correct, isn't it?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 11]
```diff
@@ -1300,6 +1300,67 @@ class ReplicaManagerTest {
...
+      assertTrue(consumerResult.assertFired.preferredReadReplica.isDefined)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: Could we also verify that the fetch purgatory was not touched? I want to verify that no delayed fetch was inserted.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 12]
```diff
@@ -1300,6 +1300,67 @@ class ReplicaManagerTest {
...
+          .setIsr(brokerList)
+          .setZkVersion(0)
+          .setReplicas(brokerList)
+          .setIsNew(false)).asJava,
+        Collections.singletonMap(topic, topicId),
+        Set(new Node(0, "host1", 0), new Node(1, "host2", 1)).asJava).build()
+      replicaManager.becomeLeaderOrFollower(1, leaderAndIsrRequest2, (_, _) => ())
+      replicaManager.getPartitionOrException(tp0).updateFollowerFetchState(1, new LogOffsetMetadata(0), 0, 0, 0)
+
+      val metadata: ClientMetadata = new DefaultClientMetadata("rack-a", "client-id",
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: Is specifying `ClientMetadata` required? We usually omit the types in Scala unless required.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 13]
```diff
@@ -1300,6 +1300,67 @@ class ReplicaManagerTest {
...
+        InetAddress.getByName("localhost"), KafkaPrincipal.ANONYMOUS, "default")
+
+      val consumerResult = fetchAsConsumer(replicaManager, tidp0,
+        new PartitionData(Uuid.ZERO_UUID, 0, 0, 100000, Optional.empty()), minBytes = 1,
+        clientMetadata = Some(metadata))
+
+      // Fetch from follower succeeds
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: I suppose that the comment is also not correct here, isn't it?

**Review by dajac — COMMENTED** [review]
@bozhao12 Thanks for the update. Left a few more comments.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 6 above)
Comment (bozhao12) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: @dajac  Thanks for your review，I made changes for the unit test and added  some  necessary  comment.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 9 above)
Comment (bozhao12) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: @dajac done

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 14]
```diff
@@ -1300,6 +1300,62 @@ class ReplicaManagerTest {
     TestUtils.assertNoNonDaemonThreads(this.getClass.getName)
   }
 
+  @Test
+  def testHasPreferredReplica(): Unit = {
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: Would `testFetchShouldReturnImmediatelyWhenPreferredReadReplicaIsDefined` be a better name?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 15]
```diff
@@ -1300,6 +1300,62 @@ class ReplicaManagerTest {
...
+      val tidp0 = new TopicIdPartition(topicId, tp0)
+
+      initializeLogAndTopicId(replicaManager, tp0, topicId)
+      when(replicaManager.metadataCache.getPartitionReplicaEndpoints(
+        any[TopicPartition], any[ListenerName])).
+        thenReturn(Map(leaderBrokerId -> new Node(leaderBrokerId, "host1", 9092, "rack-a"),
+          followerBrokerId -> new Node(followerBrokerId, "host2", 9092, "rack-b")).toMap)
+
+      // Make this replica the leader
+      val leaderAndIsrRequest2 = new LeaderAndIsrRequest.Builder(ApiKeys.LEADER_AND_ISR.latestVersion, 0, 0, brokerEpoch,
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: `leaderAndIsrRequest2` -> `leaderAndIsrRequest`?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 16]
```diff
@@ -1300,6 +1300,62 @@ class ReplicaManagerTest {
...
+          .setReplicas(brokerList)
+          .setIsNew(false)).asJava,
+        Collections.singletonMap(topic, topicId),
+        Set(new Node(0, "host1", 0), new Node(1, "host2", 1)).asJava).build()
+      replicaManager.becomeLeaderOrFollower(1, leaderAndIsrRequest2, (_, _) => ())
+      // Avoid the replica selector ignore the follower replica if it not have the data that need to fetch
+      replicaManager.getPartitionOrException(tp0).updateFollowerFetchState(followerBrokerId, new LogOffsetMetadata(0), 0, 0, 0)
+
+      val metadata = new DefaultClientMetadata("rack-b", "client-id",
+        InetAddress.getByName("localhost"), KafkaPrincipal.ANONYMOUS, "default")
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: `InetAddress.getByName("localhost")` -> `InetAddress.getLocalHost`?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 17]
```diff
@@ -1300,6 +1300,62 @@ class ReplicaManagerTest {
...
+
+      //If a preferred read replica is selected, the fetch response will return immediately，even with minBytes and timeout configured
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: * nit: Could we add a space after `//` to be consistent? 
* nit: I would rephrase the comment as follow: `If a preferred read replica is selected, the fetch response returns immediately, even if min bytes and timeout conditions are not met.`. What do you think?

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 18]
```diff
@@ -1300,6 +1300,62 @@ class ReplicaManagerTest {
...
+        new PartitionData(Uuid.ZERO_UUID, 0, 0, 100000, Optional.empty()), minBytes = 1,
+        clientMetadata = Some(metadata), timeout = 5000)
+
+      // Fetch from leader succeeds
+      assertTrue(consumerResult.isFired)
+
+      // No delayed fetch was inserted
+      assertEquals(0, replicaManager.delayedFetchPurgatory.watched)
+
+      // Returns a preferred replica,not None
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: I think that we can remove `,not None` here.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala [hunk 19]
```diff
@@ -1300,6 +1300,62 @@ class ReplicaManagerTest {
...
+      val brokerList = Seq[Integer](leaderBrokerId, followerBrokerId).asJava
+      val topicId = Uuid.randomUuid()
+      val tp0 = new TopicPartition(topic, 0)
+      val tidp0 = new TopicIdPartition(topicId, tp0)
+
+      initializeLogAndTopicId(replicaManager, tp0, topicId)
+      when(replicaManager.metadataCache.getPartitionReplicaEndpoints(
+        any[TopicPartition], any[ListenerName])).
+        thenReturn(Map(leaderBrokerId -> new Node(leaderBrokerId, "host1", 9092, "rack-a"),
+          followerBrokerId -> new Node(followerBrokerId, "host2", 9092, "rack-b")).toMap)
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: nit: I find this block hard to read. I usually tend to format `when`/`then` as follow:

```scala
when(replicaManager.metadataCache.getPartitionReplicaEndpoints(
  any[TopicPartition],
  any[ListenerName])
).thenReturn(Map(
  leaderBrokerId -> new Node(leaderBrokerId, "host1", 9092, "rack-a"),
  followerBrokerId -> new Node(followerBrokerId, "host2", 9092, "rack-b")
).toMap)
```

Could we also replace the two `any`s with the expected values?

**Review by dajac — COMMENTED** [review]
@bozhao12 Thanks for the update. I left a few nits below. We should be good to go when they are addressed.

**Comment by dajac** [comment]
@bozhao12 Could you also update the description? Thanks.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala (hunk 19 above)
Comment (bozhao12) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala]: @dajac  It looks better，all codes mentioned in the comments have been updated. Thanks.

**Review by dajac — APPROVED** [review]
LGTM, thanks for the patch! I will merge it when we get a clean build.

**Comment by ijuma** [comment]
@dajac are you aware of previous changes in this area? cc @hachikuji

**Comment by dajac** [comment]
@ijuma I am aware of some but I can't claim to be aware of all :). This seems to be a miss in the original implementation. I can't think of any reason not to complete the fetch in this case.

**Comment by ijuma** [comment]
Thanks for clarifying the PR title and description. Makes sense.

**Comment by bozhao12** [comment]
@dajac  the description has also been updated to "Fetch from consumers should return immediately when preferred read replica is defined by the leader"

**Comment by dajac** [comment]
Merged to trunk and 3.2.

**Comment by dajac** [comment]
@bozhao12 Thanks for your contribution!

**Comment by bozhao12** [comment]
@dajac @ijuma  Currently, because the consumer supports pulling data from the leader replica, and also supports pulling data from the follower replica.
Therefore, Whether fetching data from the leader or the follower, we have verified the leaderEpoch in the tryComplete method.
So I think the method scalaDoc of DelayFetch.tryComplete () also needs to be updated. I push a PR  https://github.com/apache/kafka/pull/12078 ，Looking forward to your review. Thanks.
